lib/workers/repository/update/pr/code-owners.ts

```diff
@@ -5,9 +5,100 @@ import { readLocalFile } from '../../../../util/fs';
 import { getBranchFiles } from '../../../../util/git';
 import { newlineRegex, regEx } from '../../../../util/regex';
 
+interface FileOwnerRule {
+  usernames: string[];
+  pattern: string;
+  score: number;
+  match: (path: string) => boolean;
+}
+
+function extractOwnersFromLine(line: string): FileOwnerRule {
+  const [pattern, ...usernames] = line.split(regEx(/\s+/));
+  const matchPattern = ignore().add(pattern);
+  return {
+    usernames,
+    pattern,
+    score: pattern.length,
+    match: (path: string) => matchPattern.ignores(path),
+  };
+}
+
+interface UserScore {
+  username: string;
+  score: number;
+}
+
+interface FileScore {
+  file: string;
+  score: number;
+}
+
+interface FileOwnersScore {
+  file: string;
+  usernames: UserScore[];
+}
+
+function matchFileToOwners(
+  file: string,
+  rules: FileOwnerRule[]
+): FileOwnersScore {
+  const usersWithScore = rules
+    .map((rule) =>
+      rule.match(file) ? { score: rule.score, usernames: rule.usernames } : null
+    )
+    .reduce<UserScore[]>((acc, match) => {
+      if (!match) {
+        return acc;
+      }
+      for (const user of match.usernames) {
+        if (!acc.find((a) => a.username === user)) {
+          acc.push({ score: match.score, username: user });
+        }
+      }
+
+      return acc;
+    }, []);
+
+  return { file, usernames: usersWithScore };
+}
+
+interface OwnerFileScore {
+  username: string;
+  files: FileScore[];
+}
+
+function getOwnerList(filesWithOwners: FileOwnersScore[]): OwnerFileScore[] {
+  return filesWithOwners.reduce<OwnerFileScore[]>((acc, fileMatch) => {
+    for (const userScore of fileMatch.usernames) {
+      // Get / create user file score
+      let userAcc = acc.find((u) => u.username === userScore.username);
```

Comment:
wouldn't it be better to use a map outside of the loops instead of searching in an array so often? (perf)

Reply:
wrong, renovate runs on huge repos which have hundreds of change files. so performance always matters.